File: src/main/java/fr/stardustenterprises/rust/importer/ImporterPlugin.java
package fr.stardustenterprises.rust.importer;

import org.gradle.api.NonNullApi;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.ConfigurationContainer;
import org.gradle.api.tasks.TaskContainer;

@NonNullApi
public class ImporterPlugin implements Plugin<Project> {

    @Override
    public void apply(Project project) {
        TaskContainer tasks = project.getTasks();
        ConfigurationContainer configurations = project.getConfigurations();

        Configuration configuration = configurations.create("rustImport");
        configuration.setCanBeConsumed(false);
        configuration.setCanBeResolved(true);

        ImporterTask importerTask = tasks.create("importRustResources", ImporterTask.class);
        importerTask.setGroup("rust");

        project.afterEvaluate(p -> {
            tasks.named("processResources").get().dependsOn(importerTask);
        });
    }
}

File: plugin/src/main/kotlin/fr/stardustenterprises/gradle/rust/importer/ImporterPlugin.kt
package fr.stardustenterprises.gradle.rust.importer

import fr.stardustenterprises.gradle.common.Plugin
import fr.stardustenterprises.gradle.rust.importer.ext.ImporterExtension
import org.gradle.api.Project
import org.gradle.api.artifacts.Configuration
import org.gradle.language.jvm.tasks.ProcessResources


class ImporterPlugin : Plugin() {
    override var pluginId = "fr.stardustenterprises.rust.importer"

    private lateinit var configuration: Configuration
    private lateinit var importerExtension: ImporterExtension

    override fun applyPlugin() {
        this.configuration = project.configurations.create("rust")
        this.configuration.isCanBeConsumed = false
        this.configuration.isCanBeResolved = true

        this.importerExtension = extension(ImporterExtension::class.java)
    }

    @Suppress("UnstableApiUsage")
    override fun afterEvaluate(project: Project) {
        project.tasks.withType(ProcessResources::class.java)
            .named("processResources")
            .also {
                it.configure { t ->
                    t.from(configuration)
                }
            }.get().doLast {
                val baseDir = (it as ProcessResources).destinationDir
                ProcessResourcesRust.process(project, importerExtension, baseDir)
            }
    }

    override fun conflictsWithPlugins(): Array<String> =
        arrayOf("fr.stardustenterprises.rust.wrapper")
}